adam-core/src/main/scala/org/bdgenomics/adam/serialization/BinaryDecoderWithLRUCache.scala

```diff
@@ -0,0 +1,120 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bdgenomics.adam.serialization
+
+import java.io.{ IOException, InputStream }
+import java.nio.ByteBuffer
+import java.util
+import java.util.Map.Entry
+
+import org.apache.avro.io.{ BinaryDecoder, Decoder, DecoderFactory }
+import org.apache.avro.util.Utf8
+
+/**
+ * Note: A LinkedHashMap takes 40 * SIZE + 4 * CAPACITY bytes. If the size == capacity, then the memory
+ * footprint is CAPACITY * 44 bytes.
+ */
+private[serialization] class LRUCache[K, V](maxCapacity: Int, loadFactor: Double = 0.75f)
+    extends util.LinkedHashMap[K, V](maxCapacity, 0.75f, true) {
+  assert(maxCapacity > 0, "LRU cache maxCapacity should be greater than 0")
+  assert(0 < loadFactor && loadFactor <= 1.0, "LRU load factor must be (0,1.0]")
+  override def removeEldestEntry(eldest: Entry[K, V]): Boolean = {
+    size() > maxCapacity
+  }
+}
+
+/**
+ * An Avro binary decoder that keep a LRU cache of strings to minimize memory pressure.
+ *
+ * WARNING: This class is not thread-safe! This isn't an issue since Spark guarantees that only
+ * a single thread will access it.
+ *
+ * A amount of memory strings take in Java is 8 * (int) ((((#chars) * 2) + 45) / 8) bytes, e.g. a string that is
+ * 128 characters long will take, 8 * ((((128) * 2) + 45) / 8) = 296 bytes. This BinaryDecoder is not for general
+ * use but is optimized for ADAM -omics data, where strings are typically ~128 characters long.
+ *
+ * The default cache capacity size of 64K elements, equates to a maximum memory use of (44*64K) for the cache and
+ * (296*64K) for the strings, a total of ~22 MB when full.
+ *
+ * @param in The input stream of Avro data
+ */
+private[serialization] class BinaryDecoderWithLRUCache(in: InputStream, cache: LRUCache[Utf8, Utf8]) extends Decoder {
+
+  def this(in: InputStream, cacheCapacity: Int = 65535) = this(in, new LRUCache[Utf8, Utf8](cacheCapacity))
+
+  val wrappedDecoder = DecoderFactory.get().directBinaryDecoder(in, null.asInstanceOf[BinaryDecoder])
+
+  private val scratchUtf8: Utf8 = new Utf8
+
+  @throws(classOf[IOException])
+  override def readString: String = {
+    readString(scratchUtf8).toString
+  }
+
+  @throws(classOf[IOException])
+  override def readString(old: Utf8): Utf8 = {
+    // NOTE: The 'old' parameter is intentionally being ignored. We always reuse 'scratchUtf8' to keep
+    // from creating a new Utf8 object every time we read a string. This isn't thread-safe, of course,
+    // but Spark ensures that each thread has its own serialization infrastructure.
+    val readValue = wrappedDecoder.readString(scratchUtf8)
+    if (!cache.containsKey(readValue)) {
```

Comment:
I might prefer using Guava's ```Cache<K,V>``` instead of LinkedHashMap.  The properties & behavior of the cache implementation is more configurable (see [CacheBuilder](http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/cache/CacheBuilder.html)) and I think the [get method](http://docs.guava-libraries.googlecode.com/git-history/release/javadoc/com/google/common/cache/Cache.html#get(K,%20java.util.concurrent.Callable)) semantics are more clear than the "if cached, return; otherwise create, cache and return" pattern.